src/inc/msquic_linux.h

```diff
@@ -192,76 +193,258 @@ extern char *QuicOpenSslClientTrustedCert;
+    }
+}
 
+inline
 uint16_t
 QuicAddrGetFamily(
     _In_ const QUIC_ADDR* const Addr
-    );
+    )
+{
+    return Addr->si_family;
+}
 
+inline
 void
 QuicAddrSetFamily(
     _In_ QUIC_ADDR* Addr,
     _In_ uint16_t Family
-    );
+    )
+{
+    Addr->si_family = Family;
+}
 
-uint16_t // Returns in host byte order.
+inline
+uint16_t
 QuicAddrGetPort(
     _In_ const QUIC_ADDR* const Addr
-    );
+    )
+{
+    if (AF_INET == Addr->si_family) {
+        return ntohs(Addr->Ipv4.sin_port);
+    } else {
+        return ntohs(Addr->Ipv6.sin6_port);
+    }
+}
 
+inline
 void
 QuicAddrSetPort(
-    _In_ QUIC_ADDR* Addr,
-    _In_ uint16_t Port // Host byte order
-    );
+    _Out_ QUIC_ADDR* Addr,
+    _In_ uint16_t Port
+    )
+{
+    if (AF_INET == Addr->si_family) {
+        Addr->Ipv4.sin_port = htons(Port);
+    } else {
+        Addr->Ipv6.sin6_port = htons(Port);
+    }
+}
 
+inline
 BOOLEAN
 QuicAddrIsBoundExplicitly(
     _In_ const QUIC_ADDR* const Addr
-    );
+    )
+{
+    // LINUX_TODO: How to handle IPv4? Windows just does the below.
+
+    //
+    // Scope ID of zero indicates we are sending from a connected binding.
+    //
+
+    return Addr->Ipv6.sin6_scope_id == 0;
+}
 
+inline
 void
 QuicAddrSetToLoopback(
-    _In_ QUIC_ADDR* Addr
-    );
+    _Inout_ QUIC_ADDR* Addr
+    )
+{
+    if (Addr->si_family == AF_INET) {
+        Addr->Ipv4.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    } else {
+        Addr->Ipv6.sin6_addr = in6addr_loopback;
+    }
+}
 
+inline
 uint32_t
 QuicAddrHash(
     _In_ const QUIC_ADDR* Addr
-    );
+    )
+{
+    uint32_t Hash = 5387; // A random prime number.
+#define UPDATE_HASH(byte) Hash = ((Hash << 5) - Hash) + (byte)
+    if (Addr->si_family == AF_INET) {
+        UPDATE_HASH(Addr->Ipv4.sin_port & 0xFF);
+        UPDATE_HASH(Addr->Ipv4.sin_port >> 8);
+        for (uint8_t i = 0; i < sizeof(Addr->Ipv4.sin_addr); ++i) {
+            UPDATE_HASH(((uint8_t*)&Addr->Ipv4.sin_addr)[i]);
+        }
+    } else {
+        UPDATE_HASH(Addr->Ipv6.sin6_port & 0xFF);
+        UPDATE_HASH(Addr->Ipv6.sin6_port >> 8);
+        for (uint8_t i = 0; i < sizeof(Addr->Ipv6.sin6_addr); ++i) {
+            UPDATE_HASH(((uint8_t*)&Addr->Ipv6.sin6_addr)[i]);
+        }
+    }
+    return Hash;
+}
 
+inline
 BOOLEAN
 QuicAddrIsWildCard(
     _In_ const QUIC_ADDR* const Addr
-    );
+    )
+{
+    if (Addr->si_family == AF_UNSPEC) {
+        return TRUE;
+    } else if (Addr->si_family == AF_INET) {
+        const IN_ADDR ZeroAddr = {0};
+        return memcmp(&Addr->Ipv4.sin_addr.s_addr, &ZeroAddr, sizeof(IN_ADDR)) == 0;
+    } else {
+        const IN6_ADDR ZeroAddr = {0};
+        return memcmp(&Addr->Ipv6.sin6_addr, &ZeroAddr, sizeof(IN6_ADDR)) == 0;
+    }
+}
+
+inline
+BOOLEAN
+QuicAddr4FromString(
+    _In_z_ const char* AddrStr,
+    _Out_ QUIC_ADDR* Addr
+    )
+{
+    if (AddrStr[0] == '[') {
+        return FALSE;
+    }
+    char* AddrStrCopy = (char*)malloc(strlen(AddrStr) + 1);
```

Comment:
Don't copy. Just cast.

Reply:
It's unsafe to cast. If a user passes in a string literal that's an AV

Reply:
How about this then. Don't malloc a full copy. Instead, down in the  `PortStart != NULL` scope have a stack `char[]` big enough for the maximum IPv4 address (BTW fix IPv6 func below too the same general way). Copy the address into that array and call `inet_pton` onto it. No heap allocation necessary.